### DISCUSSION TITLE
[Closed] Kotlin linting & formatting setup

### DISCUSSION THREAD

**Discussion by louwers** [maplibre/maplibre-native#679] [Polls]
Right now we use [ktlint](https://pinterest.github.io/ktlint/) for linting Kotlin files. However, this linter setup is causing issues (#675) with it not matching the IntelliJ style. I tried to fix this but actually, the authors of ktlint do not recommend using both IntelliJ and ktlint for formatting.  

For me locally the `:ktlintFormat` task is broken. And when I look at [ktlint.gradle](https://github.com/maplibre/maplibre-gl-native/blob/main/platform/android/gradle/ktlint.gradle#L22) it uses deprecated Gradle syntax and is set up in a way that is no longer supported by ktlint.

All in all, the linting / formatting setup needs to be tweaked a bit more.

- Keep using ktlint, but use one of their recommended Gradle intergrations: https://pinterest.github.io/ktlint/install/integrations/#gradle-integration
  - We should probably not use IntelliJ auto-format anymore.
  - Every time you commit you should format, maybe with a pre-commit hook so you don't forget.
- Switch to ktfmt from Facebook: https://github.com/facebook/ktfmt
  - By default set up with 2 spaces, but there is a 'Kotlin lang' code style.
  - Has an IntelliJ plugin for auto format (screenshot below) https://plugins.jetbrains.com/plugin/14912-ktfmt
  - Can be set up with spotless (formatter agnostic Gradle plugin) or [ktfmt-gradle](https://github.com/cortinico/ktfmt-gradle)

![image](https://user-images.githubusercontent.com/649392/211629498-2ad2d8b6-b1f8-40ca-891e-65276de3048b.png)

**Comment by boldtrn** [comment]
Would it be possible to use the default Android Studio Linting and Formatting?

**Reply by louwers** [reply]
I don't think there is a CLI available for that.

**Reply by boldtrn** [reply]
Ok, makes sense 👍

**Reply by artakka** [reply]
maybe another option: keep using ktlint but disable aggressive  formatting checks? Nowadays it is so easy to reformat the code to you liking it is almost irrelevant what other people use.

**Reply by boldtrn** [reply]
IMHO we need proper formatting. Ideally we mass format the whole android platform.

**Reply by boldtrn** [reply]
@artakka I think using a common formatting template helps a lot to avoid merge conflicts.

**Reply by birkskyum** [reply]
Wasn't this the cli for ktlint?

https://pinterest.github.io/ktlint/install/cli/

**Reply by louwers** [reply]
Yes, but Robin was asking about Android Studio.

**Reply by boldtrn** [reply]
Yes, I just think we should make it as easy as possible to get this right. That's why I was thinking about the Android Studio default.

For JS, there is prettier for example, this makes it very easy that everyone uses the same style.

**Comment by louwers** [comment]
I voted ktfmt, but when I tried it out it made quite significant changes to all the code. This will not help us when we looking up old commits. So I change my vote to keep using ktlint.

I found out there is an unofficial ktlint plugin for IntelliJ: https://plugins.jetbrains.com/plugin/15057-ktlint-unofficial-

I am open to @artakka's suggestion to remove formatting checks from CI.

**Reply by birkskyum** [reply]
I use this plugin. Added a hotkey to "Format with ktlint", and it's working quite well for me.

**Comment by thehoneymad** [comment]
I voted for ktlint. I think consistency is key. Linting in CLI should be consistent with the IDE.

If we do not trust CLI linting for now, I would cut an issue for that improvement and tackle that separately. And call bankruptcy on our current Gradle setup. :) To improve, sometimes we must degrade.

**Reply by louwers** [reply]
Fixed the klint Gradle intergration 👍 

https://github.com/maplibre/maplibre-gl-native/pull/685
